File: knowledge-base-backend-master/kb-user-auth/src/main/java/com/knowledge/base/userauth/service/UserService.java
package com.knowledge.base.userauth.service;

import com.baomidou.mybatisplus.core.metadata.IPage;
import com.baomidou.mybatisplus.extension.service.IService;
import com.knowledge.base.userauth.dto.RegisterDTO;
import com.knowledge.base.userauth.dto.UserDTO;
import com.knowledge.base.userauth.entity.User;
import com.knowledge.base.userauth.vo.LoginVO;
import com.knowledge.base.userauth.vo.RegisterVO;
import com.knowledge.base.userauth.vo.UserStatisticsVO;
import com.knowledge.base.userauth.vo.UserVO;

import java.util.List;

/**
 * User Service interface
 *
 * <p>Designed following the Alibaba Java Development Guidelines; provides user business logic operations</p>
 *
 * @author airwzz999
 * @since 1.0.0
 */
public interface UserService extends IService<User> {

    /**
     * User login
     *
     * @param username username
     * @param password password
     * @return login response information
     */
    LoginVO login(String username, String password);

    /**
     * Register
     *
     * @param registerDTO registration request parameters
     * @return registration response information (includes email verification status)
     */
    RegisterVO register(RegisterDTO registerDTO);

    /**
     * Verify email to activate the account
     *
     * @param token activation token
     * @return activation result message
     */
    String verifyEmail(String token);

    /**
     * User logout
     *
     * @param token token
     */
    void logout(String token);

    /**
     * Check whether a token is blacklisted (logged out)
     *
     * @param rawToken raw JWT token (without the Bearer prefix)
     * @return true if logged out/invalid, false if valid
     */
    boolean isTokenBlacklisted(String rawToken);

    /**
     * Query a user by username
     *
     * @param username username
     * @return user information
     */
    User getByUsername(String username);

    /**
     * Create a user
     *
     * @param userDTO user information
     * @return user ID
     */
    Long createUser(UserDTO userDTO);

    /**
     * Update a user
     *
     * @param userDTO user information
     * @return whether successful
     */
    Boolean updateUser(UserDTO userDTO);

    /**
     * Delete a user
     *
     * @param userId user ID
     * @return whether successful
     */
    Boolean deleteUser(Long userId);

    /**
     * Query a user by ID
     *
     * @param userId user ID
     * @return user information
     */
    UserVO getUserById(Long userId);

    /**
     * Paginated user query
     *
     * @param current  current page
     * @param size     page size
     * @param keyword  search keyword
     * @param role     role filter
     * @param status   status filter
     * @return paginated user information
     */
    IPage<UserVO> pageUsers(Long current, Long size, String keyword, String role, Integer status);

    /**
     * Reset a user's password
     *
     * @param userId          user ID
     * @param newPassword new password
     * @return whether successful
     */
    Boolean resetPassword(Long userId, String newPassword);

    /**
     * Change a user's password
     *
     * @param oldPassword old password
     * @param newPassword new password
     * @return whether successful
     */
    Boolean changePassword(String oldPassword, String newPassword);

    /**
     * Get the current logged-in user's information
     *
     * @return user information
     */
    UserVO getCurrentUserInfo();

    /**
     * Assign roles to a user
     *
     * @param userId  user ID
     * @param roleIds role ID list
     * @return whether successful
     */
    Boolean assignRoles(Long userId, List<Long> roleIds);

    /**
     * Get a user's assigned roles
     *
     * @param userId user ID
     * @return role ID list
     */
    List<Long> getUserRoles(Long userId);

    /**
     * Assign permissions to a user (directly, not via a role)
     *
     * @param userId        user ID
     * @param permissionIds permission ID list
     * @return whether successful
     */
    Boolean assignPermissions(Long userId, List<Long> permissionIds);

    /**
     * Get all of a user's permissions (including role-based and directly assigned permissions)
     *
     * @param userId user ID
     * @return permission code list
     */
    List<String> getUserPermissions(Long userId);

    /**
     * Refresh token
     *
     * @param refreshToken refresh token
     * @return new login response information (includes new access token and refresh token)
     */
    LoginVO refreshToken(String refreshToken);

    /**
     * Get a user's statistics (document count, views, likes, comments)
     *
     * @param userId user ID
     * @return user statistics
     */
    UserStatisticsVO getUserStatistics(Long userId);

    /**
     * Send a password reset verification code
     *
     * @param email registered email
     */
    void sendResetCode(String email);

    /**
     * Verify the password reset code
     *
     * @param email registered email
     * @param code  verification code
     * @return verification result
     */
    boolean verifyResetCode(String email, String code);

    /**
     * Reset password
     *
     * @param email       registered email
     * @param code        verification code
     * @param newPassword new password
     */
    void resetPassword(String email, String code, String newPassword);

    /**
     * Validate a token and return the user identity information
     *
     * @param authorization Authorization request header (Bearer token)
     * @param token         token from URL parameter (fallback)
     * @return token validation result
     */
    com.knowledge.base.userauth.vo.TokenValidateVO validateToken(String authorization, String token);

    /**
     * Query user IDs by role code
     *
     * @param roleCode role code, e.g. ROLE_REVIEWER
     * @return user ID list
     */
    List<Long> getUserIdsByRoleCode(String roleCode);
}

File: knowledge-base-frontend-master/src/services/user.service.ts
import { http } from './request';
import { User, PageParams } from '@/types';

export interface UserFilter {
  keyword?: string;
  department?: string;
  position?: string;
  role?: string;
  status?: 'active' | 'inactive';
  page?: number;
  pageSize?: number;
}

export interface UserProfileUpdate {
  username?: string;
  email?: string;
  avatar?: string;
  department?: string;
  position?: string;
  phone?: string;
  bio?: string;
}

export interface PasswordChange {
  oldPassword: string;
  newPassword: string;
  confirmPassword: string;
}

export interface PageData<T> {
  list: T[];
  total: number;
  page: number;
  pageSize: number;
}

export const userService = {
  // Get user list
  getUsers: async (filter: UserFilter): Promise<PageData<User>> => {
    const response = await http.get<any>('/auth/users/page', {
      params: {
        current: filter.page || 1,
        size: filter.pageSize || 10,
        keyword: filter.keyword,
        role: filter.role,
        status: filter.status,
      },
    });
    return {
      list: response?.records || [],
      total: response?.total || 0,
      page: response?.current || 1,
      pageSize: response?.size || 10,
    };
  },

  // Get user details
  getUser: (id: string) => {
    return http.get<User>(`/auth/users/${id}`);
  },

  // Get current user info
  getCurrentUser: () => {
    return http.get<User>('/auth/auth/me');
  },

  // Create user
  createUser: (data: {
    username: string;
    email: string;
    password?: string;
    realName?: string;
    department?: string;
    position?: string;
    status?: number;
  }) => {
    return http.post<number>('/auth/users', data);
  },

  // Update user info
  updateUser: (id: string, data: Partial<User>) => {
    return http.put<boolean>('/auth/users', { id, ...data });
  },

  // Update profile
  updateProfile: (data: UserProfileUpdate) => {
    return http.put<User>('/auth/users/me/profile', data);
  },

  // Change password
  changePassword: (data: PasswordChange) => {
    return http.put<void>('/auth/users/password/change', null, {
      params: {
        oldPassword: data.oldPassword,
        newPassword: data.newPassword,
      },
    });
  },

  // Upload avatar
  uploadAvatar: (file: File) => {
    const formData = new FormData();
    formData.append('file', file);
    return http.post<{ url: string }>('/auth/users/me/avatar', formData);
  },

  // Delete user
  deleteUser: (id: string) => {
    return http.delete<boolean>(`/auth/users/${id}`);
  },

  // Batch delete users
  batchDeleteUsers: (ids: string[]) => {
    return http.delete('/auth/users/batch', { data: { ids } });
  },

  // Enable/disable user
  toggleUserStatus: (id: string, status: 'active' | 'inactive') => {
    return http.patch(`/auth/users/${id}/status`, { status });
  },

  // Reset user password
  resetPassword: (id: string, newPassword: string) => {
    return http.put<boolean>(`/auth/users/${id}/password/reset`, null, {
      params: { newPassword },
    });
  },

  // Get user statistics
  getUserStats: (userId?: string) => {
    return http.get<{
      documentCount: number;
      viewCount: number;
      likeCount: number;
      commentCount: number;
    }>(userId ? `/auth/users/${userId}/stats` : '/auth/users/me/stats');
  },

  // Get user activity
  getUserActivity: (userId?: string, params?: PageParams) => {
    return http.get<{
      list: Array<{ type: string; description: string; createdAt: string }>;
      total: number;
    }>(userId ? `/auth/users/${userId}/activity` : '/auth/users/me/activity', { params });
  },

  // Search users
  searchUsers: (keyword: string) => {
    return http.get<User[]>('/auth/users/search', { params: { keyword } });
  },

  // Get online users
  getOnlineUsers: () => {
    return http.get<User[]>('/auth/users/online');
  },

  // Assign roles to user
  assignRoles: (userId: string, roleIds: string[], config?: any) => {
    return http.post<boolean>(
      `/auth/users/${userId}/roles`,
      roleIds.map((id) => String(id)),
      config
    );
  },

  // Get user role list
  getUserRoles: (userId: string, config?: any) => {
    return http.get<Array<string | number>>(`/auth/users/${userId}/roles`, config).then((res) =>
      (res || []).map((id) => String(id))
    );
  },

  // Assign permissions to user
  assignPermissions: (userId: string, permissionIds: string[], config?: any) => {
    return http.post<boolean>(
      `/auth/users/${userId}/permissions`,
      permissionIds.map((id) => String(id)),
      config
    );
  },

  // Get all permissions of a user
  getUserPermissions: (userId: string, config?: any) => {
    return http.get<string[]>(`/auth/users/${userId}/permissions`, config);
  },
};

export default userService;
